### PR TITLE
[top, fpga] Add support for ChipWhisperer CW310 FPGA board

### PR DESCRIPTION
**Pull request by vogelpi**
This PR adds support for the ChipWhisperer CW310 FPGA board. It's still work in progess and thus marked as draft.

The bitstream can be generated using:
```
fusesoc --cores-root . run --flag=fileset_top --target=synth lowrisc:systems:chip_earlgrey_cw310
```

Once the board is connected the cw310_loader.py utility can be used to flash the bitstream and load software images:
```
./util/fpga/cw310_loader.py -bs path_to_bitsream.bit
./util/fpga/cw310_loader.py -fw path_to_application_binary.bin
```
The UART is accessible via:
```
./util/fpga/cw310_serial.sh
```

Things that remain to be done:
- [x] Documentation:
  - [x] porting of http://cw310.com/doc/ug/getting_started_bergen/
  - ~[ ] porting of http://cw310.com/doc/ug/bergen_developer_quickstart/#ot-server-walk-through~
  - [x] adjusting Vivado installation
- [x] CI
- [ ] ...
- [ ] SCA support

### CODE REVIEW

**Comment by imphil** [comment]
Thanks Pirmin, this is a great first step!

I haven't yet looked in detail for alternatives, but on a first glance it looks like we're producing a bit too much parallel infrastructure.

* We add a rewrite of the spiflash tool specific to this board instead of extending the existing spiflash tool, or integrating it into the soon-to-be-available opentitantool (which has similar functionality). I'd like to avoid that. Let's sync up with @cfrantz and @moidx to see what the plans around spiflash are.
* I'd prefer if we can access the FPGA and do the flashing through the Xilinx JTAG port for two reasons:
	* In the CI setup we flash through the Xilinx HW server, which stays connected to the board and significantly speeds up flashing times. (And solves permission and reset sequencing issues when going through the docker container to the host.)
	* We need a working Xilinx JTAG setup for anyone wanting to do ILA FPGA debugging.
* numpy and scipy are massive Python dependencies with native code inside which are potentially painful to install for some of our users. We need to carefully look if we cannot avoid that.

**Comment by vogelpi** [comment]
Update: adjusted the Vivado installation instructions and getting started with FPGAs user guide.

Thanks for your feedback @impfhil. I have some comments to the points you raised below:

>     * We add a rewrite of the spiflash tool specific to this board instead of extending the existing spiflash tool, or integrating it into the soon-to-be-available opentitantool (which has similar functionality). I'd like to avoid that. Let's sync up with @cfrantz and @moidx to see what the plans around spiflash are.

I don't have a strong opinion about this. But I would prefer to get this PR in rather sooner than later. Now seems to be a good point in time. Once the team is again working at full blast, PRs like this will again be much more painful for everyone in the project. Adding support for the existing spiflash tool seems possible but I can't judge on the effort required for that. The main challenge is that the ChipWhisperer board doesn't have FTDI chips on it. Instead, USB connects to a single SAM3X MCU on the board which can then handle multiple UARTs as well as SPI. I don't know if that can be easily integrated into the existing `spiflash`.

The `cw_spiflash.py` utility added with this commit is not a new creation but instead moved over from the lowRISC/ot-sca repo. We have been using that already for the CW305 board. Eventually we don't want to keep maintaining two versions and the one in ot-sca can be removed.

>     * I'd prefer if we can access the FPGA and do the flashing through the Xilinx JTAG port for two reasons:
>       
>       * In the CI setup we flash through the Xilinx HW server, which stays connected to the board and significantly speeds up flashing times. (And solves permission and reset sequencing issues when going through the docker container to the host.)
>       * We need a working Xilinx JTAG setup for anyone wanting to do ILA FPGA debugging.

The main reason for not using the Xilinx JTAG port is that IIRC the CW310 board doesn't provide JTAG via FTDI/USB. There is a 10-pin Xilinx JTAG header but this requires a Xilinx Platform Cable adapter or similar. I think it wouldn't be wise to ask everyone in using the CW310 board to buy such an adapter, too. On the other hand, doing all the tooling required for that seems to be a lot of effort if it's "just" for CI. About the flashing times, please note that the new FPGA is significantly bigger than the A200T on the NexysVideo. Flashing times are already longer just because of that (now in the order of 15 seconds). But I agree with you regarding the ILA debugging. We will need support for that.

>     * numpy and scipy are massive Python dependencies with native code inside which are potentially painful to install for some of our users. We need to carefully look if we cannot avoid that.

I agree about this but they seems to be required for ChipWhisperer API which is now in charge handling bitstream flashing, UART and SPI connection to OpenTitan - all via the SAM3X MCU. Beyond that the ChipWhisperer API can of course do much more but (SCA, FI etc.) and that's probably where these dependencies come from.

**Comment by moidx** [comment]
I agree that we should think about removing dependencies. Can we work on landing this PR and filing issues to work on the tooling updates? 

CC: @cfrantz

**Comment by vogelpi** [comment]
Yes, I would prefer landing this PR and filing issues to work on the tooling. I can do that.

Update: I also extended the pytest infrastructure and enabled CI checks for the new board. In parallel, @imphil has been working on connecting a board to the server in Cambridge for running the tests.

File: util/fpga/cw_spiflash.py
Comment (mcy) [inline util/fpga/cw_spiflash.py:1]: @cfrantz This looks prime for ottoolization!

File: util/fpga/cw_spiflash.py
Comment (mcy) [inline util/fpga/cw_spiflash.py:11]: Is there a reason almost every symbol is _ prefixed...?

File: util/fpga/cw_spiflash.py
Comment (mcy) [inline util/fpga/cw_spiflash.py:33]: These are only used a couple of times. Inline them?

File: util/fpga/cw_spiflash.py
Comment (mcy) [inline util/fpga/cw_spiflash.py:261]: This else block seems unnecessary?

**Comment by imphil** [comment]
Just a quick note: I marked this ready for review mainly to get some CI checks to run. I have still some things to look into here and will do so on Monday.

File: util/fpga/cw_spiflash.py
Comment (mcy) [inline util/fpga/cw_spiflash.py:248]: Why not replace this with
```python
if !prev_frame or prev_frame.expected_ack == ack: break
```
to reduce nesting?

File: util/fpga/cw_spiflash.py
Comment (mcy) [inline util/fpga/cw_spiflash.py:211]: This hardly has any state. Maybe run() should be a free function...?

File: util/fpga/cw310_loader.py
Comment (mcy) [inline util/fpga/cw310_loader.py:17]: What is a "bergparser"...?

File: util/fpga/cw310_loader.py
Comment (mcy) [inline util/fpga/cw310_loader.py:15]: You never use argv, do you?

File: util/fpga/cw_spiflash.py
Comment (cfrantz) [reply util/fpga/cw_spiflash.py:1]: This appears to be a python implementation of the current bootstrap protocol.

I have no specific objection to this except to mention that there is already a C++ based version of the bootstrap protocol in sw/host/spiflash with some abstraction over the SPI interface such that support for different hardware interfaces (ie: FTDI chips/whatever) can be added.

Could that pre-existing code work for you?

There is a longer-term plan to unify the hardware debug interfaces to FPGA boards under a single tool, however that effort is not underway yet (and shouldn't block progress on this PR).

File: hw/top_earlgrey/chip_earlgrey_cw310.core
Comment (msfschaffner) [inline hw/top_earlgrey/chip_earlgrey_cw310.core:13]: do we still need this script for the CW310?

**Review by msfschaffner — COMMENTED** [review]
Thanks for taking a stab at this, @vogelpi and @imphil!
Topgen changes look good to me so far. 
I'll take another look once this is stable.

File: util/fpga/cw_spiflash.py
Comment (vogelpi) [reply util/fpga/cw_spiflash.py:1]: I can't judge on the feasibility, required effort of adding support for the ChipWhisperer boards to `sw/host`spiflash`. The main difference is that the CW boards don't have an FTDI but a SAM3X/U that handles SPI communication with OpenTitan.

In the early days of our SCA experiments we have been using a separate FTDI cable together with `sw/host/spiflash` for loading binaries into OpenTitan on the CW305 to avoid going through the SAM3U. It worked but the FTDI cable injected heavy noise into the power traces (even when idling) which is very bad for the experiments. Thankfully @alphan went ahead and routed SPI through the SAM3X/U to get rid of these issues. ChipWhisperer provides the Python library to handle SPI over the SAM3X/U, don't know if something similar is available in C++. Maybe @alphan can quickly comment.

Anyway, I will add a point to the tracking issue to investigate that.

**Review by vogelpi — COMMENTED** [review]
Thanks for your input guys. I've rebased this PR on current master.

File: util/fpga/cw310_loader.py
Comment (vogelpi) [reply util/fpga/cw310_loader.py]: Good catch, changed the variable name to `parser`. It was a remainder of the original script (the board is also called Bergen board).

File: util/fpga/cw310_loader.py
Comment (vogelpi) [reply util/fpga/cw310_loader.py]: Yep, good catch. Removed this.

File: .gitignore
Comment (imphil) [inline .gitignore:74]: Remove this exclude please, it's way too generic (and unnecessary once we fix the editable install, see below).

File: doc/ug/getting_started_fpga.md
Comment (imphil) [inline doc/ug/getting_started_fpga.md:81]: This guidance is misleading. Let's remove them and only point to the udev rules.

Even if you have access to the /dev/tty* device file, the chipwhisperer tooling will still fail with a beautiful chain of exceptions around "langid". I didn't fully debug it ultimately, but the user needs permissions not only to the TTY devices (listed here), but to the underlying USB devices (i.e. one layer down in the driver hierarchy), because they are used for probing for connected devices.

File: doc/ug/getting_started_fpga.md
Comment (imphil) [inline doc/ug/getting_started_fpga.md:122]: Not that it matters any more, but the recommended way of flashing for Nexys Video is the ot-pgm-fpga tool.

File: python-requirements.txt
Comment (imphil) [inline python-requirements.txt:18]: Remove this

File: python-requirements.txt
Comment (imphil) [inline python-requirements.txt:28]: ... and this. They are dependencies of the chipwhisperer package and installed through it.

File: python-requirements.txt
Comment (imphil) [inline python-requirements.txt:39]: I didn't find a reason to install chipwhipsperer in editable mode, the loading seems to work without it? If we don't have one, let's avoid it to give us a cleaner installation.

The chipwhisperer git archive is rather large, installing from an archive makes the installation much faster (it's still a 60 MB download!).

```suggestion
# Development version of ChipWhisperer toolchain
# Use a development version until support for the CW310 board works in a
# released version.
# Prefer the archive download over a git clone to decrease installation time
# (The chipwhisperer repository is rather large and uses additionally uses
# submodules, which need to be fetched as well.)
https://github.com/newaetech/chipwhisperer/archive/9b6825b495f14f85aae8b11007c63c59a1654584.tar.gz
```

File: site/docs/config.toml
Comment (imphil) [inline site/docs/config.toml:55]: No need for that either without the modified chipwhisperer package installation.

File: test/systemtest/utils.py
Comment (imphil) [inline test/systemtest/utils.py:370]: This function is now modified to work with both spiflash tool variants by assuming some uniformity in the command line arguments between the two tools, and passing others in through spiflash_bin_arg. I find that a bit too fragile.

Let's either have two functions load_sw_over_spi_with_xxx(), or add a argument that explicitly chooses one or the other tool variant.

(Or make the command line of both tools really identical to make them drop-in replacements.)

File: util/fpga/cw310_serial.sh
Comment (imphil) [inline util/fpga/cw310_serial.sh:6]: Do we really need this script? We don't have it for other boards.

**Review by imphil — COMMENTED** [review]
Thanks Pirmin for your work on this bringup.

I added a couple comments that I think need to be resolved before merging. 

There are some things I would prefer to see done before merging (reduce spiflash duplication, replacement of Nexys Video for earl grey), but I'm fine with doing them as follow-ups as well.

**Comment by imphil** [comment]
And flake8 is reporting lint warnings, can you have a look at them as well?


```
  flake8 test/systemtest/config.py test/systemtest/conftest.py test/systemtest/earlgrey/test_fpga_cw310.py test/systemtest/earlgrey/test_fpga_nexysvideo.py test/systemtest/utils.py util/fpga/cw310_loader.py util/fpga/cw_spiflash.py
	test/systemtest/config.py:63:5: E265 block comment should start with '# '
	test/systemtest/config.py:65:5: E265 block comment should start with '# '
	test/systemtest/utils.py:515:101: E501 line too long (125 > 100 characters)
```

**Comment by imphil** [comment]
The CW310 synthesis completed in CI and the tests ran on the FPGA. At least one test is failing currently, the `dif_gpio_smoketest`:

```
2021-05-24T11:06:56.5071676Z DEBUG    root:utils.py:363 E00000 dif_gpio_smoketest.c:57] CHECK-fail: AAAA != AA02
```

File: util/fpga/cw_spiflash.py
Comment (alphan) [reply util/fpga/cw_spiflash.py:11]: Mostly because we tend to do `import foo` and call something from `foo` quite a bit in the ot-sca repo and things other than `SAM3UProgrammer` and `FtdiProgrammer` were not meant to be used like that.

File: util/fpga/cw_spiflash.py
Comment (alphan) [reply util/fpga/cw_spiflash.py:33]: IIRC the main reason was just to make the f-string below shorter (`_le_to_uint`), `_uint_to_le` just for consistency.

File: util/fpga/cw_spiflash.py
Comment (alphan) [reply util/fpga/cw_spiflash.py:261]: Just a matter of preference. Yes it doesn't do anything because of `pass`, exists just for clarity because there are multiple things nested.

File: util/fpga/cw_spiflash.py
Comment (alphan) [reply util/fpga/cw_spiflash.py:211]: When this was initially implemented, the `OpenTitan` class expected the programmer to have a `run` method and this class was written to conform to that interface.

File: util/fpga/cw310_loader.py
Comment (vogelpi) [reply util/fpga/cw310_loader.py]: Done.

File: .gitignore
Comment (vogelpi) [reply .gitignore]: Done.

File: doc/ug/getting_started_fpga.md
Comment (vogelpi) [reply doc/ug/getting_started_fpga.md]: That's changed now as well. I don't remember exactly but I thought I had these problems as well but they went away after re-plugging the boards once udev rules are set up.

File: doc/ug/getting_started_fpga.md
Comment (vogelpi) [reply doc/ug/getting_started_fpga.md]: I think it does matter. I've changed the instructions to refer to the `opentitan-pgm-fpga` tool. Thanks for pointing this out.

File: python-requirements.txt
Comment (vogelpi) [reply python-requirements.txt]: Done.

File: python-requirements.txt
Comment (vogelpi) [reply python-requirements.txt]: Done as well. The reason for adding them was ChipWhisperer mentioning that sometimes installation/compilation of these packages fails when using chipwhisperer in editable mode.

File: util/fpga/cw_spiflash.py
Comment (alphan) [reply util/fpga/cw_spiflash.py:1]: As @vogelpi mentioned, this implements the current bootstrap protocol using SAM3U's [gpio mode](https://chipwhisperer.readthedocs.io/en/latest/api.html#cw305-sam3u-io-control), i.e. `host -> SAM3U -> CW305` instead of `host -> ftdi -> CW305`. 

That being said, I'm not sure if it's a good idea to keep this script in the opentitan repo since it only exists to reduce noise. It would be great if we could use CW310 in opentitan using both ROMs, i.e. `boot_rom` and `mask_rom`, and existing tools and keep this script in ot-sca. I'm not sure if we will have issues with pins though. @vogelpi WDYT?

File: python-requirements.txt
Comment (vogelpi) [reply python-requirements.txt]: Thanks, this is clearly the better option. I've implemented the change as suggested and also tested it successfully.

About the reason for the editable mode: It's actually recommend by ChipWhisperer: https://chipwhisperer.readthedocs.io/en/latest/installing.html#chipwhisperer

But as the method you suggested works as well, I think it's better to stick to this for the moment.

File: site/docs/config.toml
Comment (vogelpi) [reply site/docs/config.toml]: Done.

File: util/fpga/cw310_serial.sh
Comment (vogelpi) [reply util/fpga/cw310_serial.sh]: Removed.

File: test/systemtest/utils.py
Comment (vogelpi) [reply test/systemtest/utils.py]: Yes, I agree. This isn't ideal. I've now changed the implementation to generate the required argument based on `spiflash_path`. This way, the calling function just specifies the loader tool and the binary (+ optional arguments). The rest is handled by `load_sw_over_spi()`. Having another argument for the tool selection could lead to even more confusion.

**Comment by vogelpi** [comment]
> And flake8 is reporting lint warnings, can you have a look at them as well?
> 
> ```
>   flake8 test/systemtest/config.py test/systemtest/conftest.py test/systemtest/earlgrey/test_fpga_cw310.py test/systemtest/earlgrey/test_fpga_nexysvideo.py test/systemtest/utils.py util/fpga/cw310_loader.py util/fpga/cw_spiflash.py
> 	test/systemtest/config.py:63:5: E265 block comment should start with '# '
> 	test/systemtest/config.py:65:5: E265 block comment should start with '# '
> 	test/systemtest/utils.py:515:101: E501 line too long (125 > 100 characters)
> ```

Done.

**Comment by vogelpi** [comment]
> The CW310 synthesis completed in CI and the tests ran on the FPGA. At least one test is failing currently, the `dif_gpio_smoketest`:
> 
> ```
> 2021-05-24T11:06:56.5071676Z DEBUG    root:utils.py:363 E00000 dif_gpio_smoketest.c:57] CHECK-fail: AAAA != AA02
> ```

Looking into that.

File: test/systemtest/utils.py
Comment (imphil) [inline test/systemtest/utils.py:520]: This line is intentionally long to be exactly what we are trying to match in the pattern below. You can add noqa comments to it if lint complains.

**Review by imphil — APPROVED** [review]
LGTM as first step once CI passes. Thanks @vogelpi!

File: util/fpga/cw_spiflash.py
Comment (alphan) [inline util/fpga/cw_spiflash.py:148]: I think this should be a different mapping defined above (maybe using namedtuples for each board), and check that board is either CW310 or CW305.

File: util/fpga/cw_spiflash.py
Comment (alphan) [reply util/fpga/cw_spiflash.py:1]: > I can't judge on the feasibility, required effort of adding support for the ChipWhisperer boards to `spiflash`. The main difference is that the CW boards don't have an FTDI but a SAM3X/U that handles SPI communication with OpenTitan.

`spiflash` already supports programming using an FTDI cable (of course, CW310's pins are routed differently now to be able to use SAM3U) so that can be an option using existing code and hardware. Removing the FTDI cable requires talking to CW310 endpoints over USB as this script does. This script uses `chipwhisperer` (which we need anyway for loading bitstreams, configuring CW310, and occasionally updating SAM3U's firmware) to do that. 

> ChipWhisperer provides the Python library to handle SPI over the SAM3X/U, don't know if something similar is available in C++.

I don't think so.

On a second thought, it may not be a bad idea to use this script in opentitan too since it doesn't require an FTDI cable and simplifies our setup. It would be great if we could enhance this to support the new bootstrap protocol that's in the works.

File: test/systemtest/utils.py
Comment (vogelpi) [reply test/systemtest/utils.py]: Done.

File: python-requirements.txt
Comment (colinoflynn) [reply python-requirements.txt]: For posterity - the editable mode was recommended as many users are tracking changes in CW rather than using specific releases, and end up with "local changes" themselves. So it was more of a usability feature we found, but forgot to mention the reason in our docs. So will fix that (doesn't affect OT).

**Comment by colinoflynn** [comment]
I didn't see the PR before sorry, so didn't realize some of the issues raised I might be able to give some insight on:

File: util/fpga/cw310_serial.sh
Comment (colinoflynn) [reply util/fpga/cw310_serial.sh]: When we were testing, the serial port would block if not using `cs8,-ixon,-ixoff` so had added that script out of laziness. We didn't normally use `screen` as our serial terminal interface of choice however, so wasn't sure if that was somehow related to our system or not. All the docs etc used screen as that seemed to be the default in OpenTitan docs.

**Comment by colinoflynn** [comment]
I didn't see this PR on main repo until now - might have some minor insight helpful here:

* On numpy & scipy: this is indeed part of the "total ChipWhisperer experience", but wouldn't be needed for full support. I'll add details to #6683.

* On USB interface C/C++: We don't have C/C++ interface for this, the protocol is pretty basic so it could be done (it talks via libusb which has good C/C++ support). I'm personally on the Python train however so we haven't supported this before, as we have found the pyusb interface works well for cross-platform support (a lot of development is on Windows even here...).

* On USB JTAG: The "Xilinx JTAG over FTDI" is a magic Digilent option which came out of some licensing or other agreement from what I know. There has been some effort that R.E.'d that, but it's never been clear someone else can use it w/o Xilinx's blessing, at least in a released product. Basically there is some "magic sequence" in the FTDI eeprom to enable this, but I doubt you could add it to anything FOSS. Relative to the cost of the board (or even the Kintex FPGA) the cost of either a Platform USB cable is not substantial, so we just pined out the standard header here.

* On USB/CI: In theory the SAM3X USB is stable so you don't need to disconnect/reconnect - that is the USB connection stays alive between CI runs. But in case of firmware issues it could be useful, we had previously setup a docker container as part of automated testing that would pass-through our USB interfaces and seemed stable with this.

**Comment by vogelpi** [comment]
Thanks @colinoflynn for your input!

In the meantime, I have also generated some area reports and timing summaries for the current design (`reduced`: still using the hacky reduce script to keep binary compatibility with NexysVideo) and the design with full SRAM and Flash sizes (we will soon move to that config). As you can see from the reports below, all IPs that are currently in the top-level are actually implemented (including OTBN and entropy source that were previously stubbed out on the NexysVideo) and AES also has the masking enabled. With that config, we currently use 54% of LUTs, and 55% of BRAM (34% for reduced version).

[top_earlgrey_reduced_utilization_hierarchical.txt](https://github.com/lowRISC/opentitan/files/6540330/top_earlgrey_reduced_utilization_hierarchical.txt)
[top_earlgrey_utilization_hierarchical.txt](https://github.com/lowRISC/opentitan/files/6540331/top_earlgrey_utilization_hierarchical.txt)

[top_earlgrey_reduced_timing_summary.txt](https://github.com/lowRISC/opentitan/files/6540338/top_earlgrey_reduced_timing_summary.txt)
[top_earlgrey_timing_summary.txt](https://github.com/lowRISC/opentitan/files/6540345/top_earlgrey_timing_summary.txt)

**Comment by msfschaffner** [comment]
Awesome, thanks for getting this one landed!!

**Pull request merged**
